feat: introduce RunnableDevEnvironment (#18190)
vitejs/vite: docs/guide/api-environment.md

}
```

> Vite also supports a `RunnableDevEnvironment`, which is a subset of a `DevEnvironment` that exposes a `ModuleRunner` instance. You can guard any runnable environment with an `isRunnableDevEnvironment` function.

Reviewer: ```suggestion
Vite also supports a `RunnableDevEnvironment`, that extends a `DevEnvironment` exposing a `ModuleRunner` instance. You can guard any runnable environment with an `isRunnableDevEnvironment` function.
```

:::warning
The `runner` is evaluated eagerly when it's accessed for the first time. Beware that Vite enables source map support when the `runner` is created by calling `process.setSourceMapsEnabled` or by overriding `Error.prepareStackTrace` if it's not available.
:::

```ts
export class RunnableDevEnvironment extends DevEnvironment {
  public readonly runner: ModuleRunnner
}

if (isRunnableDevEnvironment(server.environments.ssr)) {
  await server.environments.ssr.runner.import('/entry-point.js')

> }

Reviewer: In the same way that we have `environment.mode: 'dev' | 'build'`, should we have `environment.type: 'runnable' | 'fetchable' | 'custom'`?

We can move forward with the function and iterate later though. Maybe the function is fine. If we decide for it, I wonder if we should expose `isDevEnvironment()` and `isBuildEnvironment()` 🤔

Author: The environment can be both runnable and fetchable; you just need a `dispatchFetch` function on it

Reviewer: Ah, I forgot about that. Ok, it makes sense then.
